runtime/lua/uri.lua

```diff
@@ -0,0 +1,123 @@
+--- TODO: This is implemented only for files now.
+-- https://tools.ietf.org/html/rfc3986
+-- https://tools.ietf.org/html/rfc2732
+-- https://tools.ietf.org/html/rfc2396
+
+local URI = {}
+
+URI.__index = URI
+URI.new = function(scheme, authority, path, query, fragment)
+  local obj = setmetatable({
+      scheme = scheme,
+      authority = authority,
+      path = path,
+      query = query,
+      fragment = fragment,
+    }, URI)
+  return obj
+end
+
+URI.tostring = function(self)
+  local res = ''
+  local scheme = self.scheme
+  local authority = self.authority
+  local path = self.path
+
+  res = res..scheme..':'
+
+  if authority or scheme == 'file'then
+    if URI.is_windows_filepath(path) then
+      res = res..'///'
+    else
+      res = res..'//'
+    end
+  end
+
+  return res..path
+end
+
+URI.from_filepath = function(path)
+  if URI.is_windows_filepath(path) then
+    local volume_path = vim.api.nvim_call_function('substitute', { path, '\\(^\\c[A-Z]:\\)\\(.*\\)', '\\1', '' })
+    local file_path = vim.api.nvim_call_function('substitute', { path, '\\(^\\c[A-Z]:\\)\\(.*\\)', '\\2', '' })
+
+    path = volume_path..URI.encode(vim.api.nvim_call_function('substitute', { file_path, '\\', '/', 'g'}))
+  else
+    path = URI.encode(path)
+  end
+
+  return URI.new('file', nil, path, nil, nil)
+end
+
+URI.is_windows_filepath = function(path)
+  if not (path:find('^[A-Z]:') == nil) then
+    return true
+  end
+  return false
+end
+
+URI.is_windows_uri = function(uri)
+  if uri:gsub('^file://', ''):find('^/[A-Z]:') then
+    return true
+  end
+  return false
+end
+
+URI.from_bufnr = function(bufnr)
+  if bufnr == nil then
+    bufnr = vim.api.nvim_get_current_buf()
+  end
+
+  return URI.from_filepath(vim.api.nvim_buf_get_name(bufnr))
+end
```

Comment:
This function is unnecessary. Callers should just do it directly:

    vim.uri_from_fname(vim.api.nvim_buf_get_name(...))